lib/proc-macros/Cargo.toml

```diff
@@ -0,0 +1,15 @@
+[package]
+name = "proc-macros"
```

Comment:
All of our own crates are prefixed with `jj-`, so `jj-proc-macros`

Reply:
This is also pretty important given there's `proc-macro`, `proc-macro2`, and now `proc-macros`(!?)

Reply:
Ack regarding the `jj-` prefix, that makes sense.

I'm less sure about moving the directory. I actually don't think that this macro will work properly if used outside of the lib crate since it refers to the `ContentHash` trait by referencing the crate via `crate::content_hash::ContentHash::hash` (this is what the previous macro did as well). I think there's a way to reference it using `content_hash::ContentHash`, but it requires three crates: one for the trait, one for the macro, and one that re-exports both of them. I'm not 100% certain about this, but I think it's something along those lines.

I think if we need proc macros for the CLI they will probably be different macros and can be implemented in a cli subcrate, so there's less of a benefit to having a single `jj-proc-macros` crate.

We could also write the config.toml files in a way where the proc macros crate isn't listed as a workspace dependency in the top-level cargo.toml file (similar to how gen-protos isn't). Perhaps that's a good idea as well?

I can experiment with a few things to confirm my understanding, but let me know what you think.

Reply:
Appears that there's `extern crate self as ..` hack, but I don't know if that's common practice.
https://stackoverflow.com/a/57049687

Reply:
> I pushed a second "WIP" commit to this PR that wires things up so that the `#[derive(ContentHash)]` works both inside and outside of `jj_lib`.

The change looks good to me, thanks. I don't have expertise of proc macros, so it would be nice if someone can rubber stamp on it.

Regarding the directory structure, I think the current lib/proc-macros is good. External users should use the re-exported version instead of directly importing from jj_lib_proc_macros.

Reply:
Great. I agree this is the way to go for the the directory structure.

I've squashed the WIP commit and cleaned up a few things. There's a chain of 4 PRs now related to adding this proc macro. You should be able to see the list in #3054. It would be great to start merging these soon.

Reply:
Resolving...